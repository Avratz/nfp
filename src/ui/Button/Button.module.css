.button {
  border-radius: 5rem;
  display: inline-block;
  text-align: center;
  font-family: var(--font-dmSans);
  font-weight: var(--font-bold);
  cursor: pointer;
  border: 2px solid;
}

.primary {
  color: var(--neutral-8);
  background-color: transparent;
  border-color: var(--neutral-4);
}

.primaryDark {
  color: var(--neutral-2);
  background-color: transparent;
  border-color: var(--neutral-3);
}

.accent {
  color: var(--neutral-8);
  background-color: var(--accent);
  border-color: var(--accent);
}

.icon {
  color: var(--neutral-8);
  background-color: transparent;
  border: 0 none;
  width: 32px;
  height: 32px;
  position: relative;
  border-radius: 0;
}

.fullWidth {
  width: 100%;
}

.xs {
  font-size: var(--text-xs);
  padding: 0.25rem 0.5rem;
}

.small {
  font-size: var(--text-sm);
  padding: 0.5rem 0.75rem;
}

.base {
  font-size: var(--text-base);
  padding: 0.75rem 1rem;
}

.medium {
  font-size: var(--text-md);
  padding: 1rem 1.25rem;
}

.loading {
  opacity: 0.5;
  cursor: not-allowed;
  pointer-events: none;
  animation: loading 0.5s infinite alternate;
}

@keyframes loading {
  0% {
    opacity: 0.5;
  }
  100% {
    opacity: 0.8;
  }
}

.blocked {
  opacity: 0.5;
  cursor: not-allowed;
  pointer-events: none;
}
