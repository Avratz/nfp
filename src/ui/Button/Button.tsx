import { forwardRef } from 'react'
import { cva, VariantProps } from 'class-variance-authority'
import { motion } from 'framer-motion'

import { ButtonOrLink } from '@/ui/ButtonOrLink/ButtonOrLink'
import type { ButtonOrLinkProps } from '@/ui/ButtonOrLink/ButtonOrLink'

import styles from './Button.module.css'

const buttonStyles = cva(styles.button, {
  variants: {
    intent: {
      primary: styles.primary,
      primaryDark: styles.primaryDark,
      accent: styles.accent,
      icon: styles.icon,
    },
    fullWidth: {
      true: styles.fullWidth,
    },
    size: {
      xs: styles.xs,
      small: styles.small,
      base: styles.base,
      medium: styles.medium,
    },
    loading: {
      true: styles.loading,
    },
    blocked: {
      true: styles.blocked,
    },
  },
  defaultVariants: {
    intent: 'primary',
    size: 'base',
    loading: false,
    blocked: false,
  },
})

export interface ButtonProps
  extends ButtonOrLinkProps,
    VariantProps<typeof buttonStyles> {}

export const Button = forwardRef<
  HTMLButtonElement & HTMLAnchorElement,
  ButtonProps
>(function forwardButton(
  {
    intent,
    fullWidth,
    size,
    className,
    loading,
    blocked,
    ...props
  }: ButtonProps,
  ref,
) {
  return (
    <ButtonOrLink
      className={buttonStyles({
        intent,
        fullWidth,
        size,
        loading,
        blocked,
        className,
      })}
      {...props}
      ref={ref}
    />
  )
})

export const MotionButton = motion(Button)
