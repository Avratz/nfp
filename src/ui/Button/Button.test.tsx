import { render } from '@testing-library/react'
import { Button } from './Button'

describe('Button', () => {
  it('should render default', () => {
    const { container } = render(<Button>default</Button>)
    expect(container).toMatchSnapshot()
  })

  it('should render primary', () => {
    const { container } = render(<Button intent="primary">primary</Button>)
    expect(container).toMatchSnapshot()
  })

  it('should render accent', () => {
    const { container } = render(<Button intent="accent">accent</Button>)
    expect(container).toMatchSnapshot()
  })

  it('should render primaryDark', () => {
    const { container } = render(
      <Button intent="primaryDark">primaryDark</Button>,
    )
    expect(container).toMatchSnapshot()
  })
})
