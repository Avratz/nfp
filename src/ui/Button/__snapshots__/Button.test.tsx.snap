// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render accent 1`] = `
<div>
  <button
    class="button accent base"
  >
    accent
  </button>
</div>
`;

exports[`Button should render default 1`] = `
<div>
  <button
    class="button primary base"
  >
    default
  </button>
</div>
`;

exports[`Button should render primary 1`] = `
<div>
  <button
    class="button primary base"
  >
    primary
  </button>
</div>
`;

exports[`Button should render primaryDark 1`] = `
<div>
  <button
    class="button primaryDark base"
  >
    primaryDark
  </button>
</div>
`;
